arbor/backends/gpu/matrix_fine.cu

```diff
@@ -0,0 +1,317 @@
+/// Each block solves a set of matricesb iterating over the levels of matrix
+/// and perfoming a backward and forward substitution. On each level one thread
+/// gets assigned to one branch on this level of a matrix and solves and
+/// performs the substitution. Afterwards all threads continue on the next
+/// level.
+/// To avoid idle threads, one should try that on each level, there is a similar
+/// number of branches.
+template <typename T>
+__global__
+void solve_matrix_fine(
+    T* rhs,
+    T* d,
+    const T* u,
+    const level* levels,
+    const unsigned* levels_end,
+    unsigned* num_matrix, // number of packed matrices = number of cells
+    unsigned* padded_size)
+{
+    const auto tid = threadIdx.x;
+    const auto bid = blockIdx.x;
+
+    const auto first_level = (bid!= 0 ? levels_end[bid - 1] : 0);
+    const auto num_levels  = levels_end[bid] - first_level;
+    const auto block_levels = &levels[first_level];
+
+    // backward substitution
+
+    for (unsigned l=0; l<num_levels-1; ++l) {
+        const auto& lvl = block_levels[l];
+
+        const unsigned width = lvl.num_branches;
+
+        // Perform backward substitution for each branch on this level.
+        // One thread per branch.
+        if (tid < width) {
+            const unsigned len = lvl.lengths[tid];
+            unsigned pos = lvl.data_index + tid;
+
+            // Zero diagonal term implies dt==0; just leave rhs (for whole matrix)
+            // alone in that case.
+
+            // Each cell has a different `dt`, because we choose time step size
+            // according to when the next event is arriving at a cell. So, some
+            // cells require more time steps than others, but we have to solve
+            // all the matrices at the same time. When a cell finishes, we put a
+            // `0` on the diagonal to mark that it should not be solved for.
+            if (d[pos]!=0) {
+
+                // each branch perform substitution
+                T factor = u[pos] / d[pos];
+                for (unsigned i=0; i<len-1; ++i) {
+                    const unsigned next_pos = pos + width;
+                    d[next_pos]   -= factor * u[pos];
+                    rhs[next_pos] -= factor * rhs[pos];
+
+                    factor = u[next_pos] / d[next_pos];
+                    pos = next_pos;
+                }
+
+                // Update d and rhs at the parent node of this branch.
+                // A parent may have more than one contributing to it, so we use
+                // atomic updates to avoid races conditions.
+                const unsigned parent_index = block_levels[l+1].data_index;
+                const unsigned p = parent_index + lvl.parents[tid];
+                //d[p]   -= factor * u[pos];
+                cuda_atomic_add(d  +p, -factor*u[pos]);
+                //rhs[p] -= factor * rhs[pos];
+                cuda_atomic_add(rhs+p, -factor*rhs[pos]);
+            }
+        }
+        __syncthreads();
+    }
+
+    {
```

Comment:
I added the special case because this allows us to skip one `__syncthreads` and the atomics. When we solve the root branch, we know that there is only one active thread working on each matrix and we do not have to synchronize. 
Once the tree balancing works the root branch will have length one again and we could use something like the old code again (if the tree does not only consist of two compartments). But this version ensures that the solver also works if we do not balance the tree, or if we balance it in another way where the root branch can have a length greater than one.

Reply:
The tree balancing doesn't work yet?

Reply:
I'm quite sure that the _tree_-balancing works, but I think there is still a bug related to the solver / level information which have to be changed if the tree was changed, becuase I get wrong results when I choose a root other than 0 (which has no effect but changing the indices of the unknown if the indices in a branch are not coninuous)

Reply:
No worries. The current version is already much faster, so let's focus on merging without the tree rebalancing turned on.